source/common/http/http2/metadata_encoder.cc

```diff
@@ -79,14 +84,21 @@ bool MetadataEncoder::createHeaderBlockUsingNghttp2(const MetadataMap& metadata_
 
 bool MetadataEncoder::hasNextFrame() { return payload_.length() > 0; }
 
-uint64_t MetadataEncoder::packNextFramePayload(uint8_t* buf, const size_t len) {
-  ASSERT(!payload_size_queue_.empty());
+ssize_t MetadataEncoder::packNextFramePayload(uint8_t* buf, const size_t len) {
+  if (payload_size_queue_.empty()) {
+    ENVOY_LOG(error, "No payload remaining to pack into a METADATA frame.");
```

Comment:
Do we unit test all the (awesome!) new error handling?  I'm kicking the coverage run but if you EXPECT_LOG_CONTAINS in your unit tests it'd both regression-proof you cover the error you're aiming for and make it easy for your reviewer to tell :-)

Reply:
Here are a couple of EXPECT_LOG_CONTAINS tests. I didn't see how to hit them all without somehow mocking out the nghttp2 calls:
- We could fail to deflate the metadata payload if we run out of memory, but I don't see any other way to cause that failure.
- The payload buffer could only be too small if an update to nghttp2 changes the minimum payload length it sends us. (I don't know how safe it is to assume that won't happen.)

Reply:
The consistency issue you mentioned is one possible way this could happen. It could also happen if a bug somewhere in nghttp2 caused it to call this callback more times than we enqueued METADATA frames. The style guide says that bad data from an API call is the kind of thing we should handle gracefully instead of crashing, so I don't think RELEASE_ASSERT is the right tool here.

Reply:
I would at least leave a comment explaining why this check is here.

Reply:
Sure, done.

Reply:
In general I'm uncomfortable with this type of error logging without either a stat or just a RELEASE_ASSERT or normal ASSERT. Sorry just to clarify again, why do we think that RELEASE_ASSERT is not OK? Isn't this a bug in nghttp2? If we decide we want to handle this I think we need a stat vs. a log that could potentially spam?

Reply:
Treating nghttp2 as an API, my reading of the style guide was that we should handle gracefully instead of crashing (see above) - but since you also think a RELASE_ASSERT is better, I've just changed it back to that.

(Would have liked to hear this from you a bit earlier in the process, though - this review has been taking a while!)